tokio/src/sync/semaphore.rs

```diff
@@ -47,28 +47,63 @@ use std::sync::Arc;
-/// ```
+/// Similar to limiting the numer of simultaneous opened files, network handles are a limited resource.
+/// Allowing an unbounded amount of requests to be processed could result in a denial-of-service,
+/// and many other issues.
+///
+/// However, in contrast to the file example, this example uses a non-static, `Arc`-wrapped `Semaphore`
+/// for more fine-grained lifetime and ownership control. The `Arc` allows multiple threads and tasks
+/// to share ownership of the semaphore. Here we create a shallow-copy of the `Arc<Semaphore>` reference
+/// using `clone`. Then we acquire a permit through the `Arc` from the `Semaphore` via [`Semaphore::acquire_owned`],
+/// and move it inside the task. This ensures no non-`'static` variables are referenced from within
+/// said task.
+///
+/// If we leave the scope where the `Arc<Semaphore>` was defined, and references still exist in any
+/// of our tasks to it, the `Semaphore` will continue to exist in a static-like context, until all
+/// `Arc`s have been dropped.
+///
+/// ```no_run
 /// use std::sync::Arc;
 /// use tokio::sync::Semaphore;
+/// use tokio::net::TcpListener;
+/// use tokio::io::{AsyncReadExt, AsyncWriteExt};
 ///
 /// #[tokio::main]
-/// async fn main() {
+/// async fn main() -> Result<(), Box<dyn std::error::Error>> {
 ///     let semaphore = Arc::new(Semaphore::new(3));
-///     let mut join_handles = Vec::new();
+///     let listener = TcpListener::bind("127.0.0.1:8080").await?;
 ///
-///     for _ in 0..5 {
+///     loop {
 ///         let permit = semaphore.clone().acquire_owned().await.unwrap();
-///         join_handles.push(tokio::spawn(async move {
-///             // perform task...
-///             // explicitly own `permit` in the task
-///             drop(permit);
-///         }));
-///     }
+///         let (mut socket, _) = listener.accept().await?;
+///
+///         tokio::spawn(async move {
+///                 let mut buf = [0; 1024];
+///
+///                 // In a loop, read data from the socket and write the data back.
+///                 loop {
+///                     let n = match socket.read(&mut buf).await {
+///                         // socket closed
+///                         Ok(n) if n == 0 => break,
+///                         Ok(n) => n,
+///                         Err(e) => {
+///                             eprintln!("failed to read from socket; err = {:?}", e);
+///                             break;
+///                         }
+///                     };
+///
+///                     // Write the data back
+///                     if let Err(e) = socket.write_all(&buf[0..n]).await {
+///                         eprintln!("failed to write to socket; err = {:?}", e);
+///                         break;
+///                     }
+///                 }
 ///
-///     for handle in join_handles {
-///         handle.await.unwrap();
+///                 // Drop the permit, so more tasks can be created
+///                 drop(permit);
```

Comment:
I would drop the socket first. Otherwise, there's technically a short period where the socket is still open, even though we have released the permit.
```suggestion
///                 drop(socket);
///                 // Drop the permit, so more tasks can be created
///                 drop(permit);
```